File: bukkit/src/main/java/me/neznamy/tab/platforms/bukkit/features/PerWorldPlayerlist.java
package me.neznamy.tab.platforms.bukkit.features;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerChangedWorldEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.plugin.java.JavaPlugin;

import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.cpu.TabFeature;
import me.neznamy.tab.shared.cpu.UsageType;
import me.neznamy.tab.shared.features.types.Loadable;

/**
 * Per-world-playerlist feature handler
 */
@SuppressWarnings("deprecation")
public class PerWorldPlayerlist implements Loadable, Listener {
	
	//plugin instance
	private JavaPlugin plugin;
	
	//config options
	private boolean allowBypass;
	private List<String> ignoredWorlds;
	private Map<String, List<String>> sharedWorlds;

	/**
	 * Constructs new instance with given parameters and loads config options
	 * @param plugin - plugin instance
	 * @param tab - tab instance
	 */
	public PerWorldPlayerlist(JavaPlugin plugin, TAB tab) {
		this.plugin = plugin;
		allowBypass = tab.getConfiguration().getConfig().getBoolean("per-world-playerlist.allow-bypass-permission", false);
		ignoredWorlds = tab.getConfiguration().getConfig().getStringList("per-world-playerlist.ignore-effect-in-worlds", Arrays.asList("ignoredworld", "build"));
		sharedWorlds = tab.getConfiguration().getConfig().getConfigurationSection("per-world-playerlist.shared-playerlist-world-groups");
		for (Entry<String, List<String>> group : sharedWorlds.entrySet()) {
			if (group.getValue() == null) {
				tab.getErrorManager().startupWarn("World group \"" + group + "\" in per-world-playerlist does not contain any worlds. You can just remove the group.");
			} else if (group.getValue().size() == 1) {
				tab.getErrorManager().startupWarn("World group \"" + group + "\" in per-world-playerlist only contain a single world (\"" + group.getValue().get(0) +
						"\"), which has no effect and only makes config less readable. Delete the group entirely for a cleaner config.");
			}
		}
		tab.debug(String.format("Loaded PerWorldPlayerlist feature with parameters allowBypass=%s, ignoredWorlds=%s, sharedWorlds=%s", allowBypass, ignoredWorlds, sharedWorlds));
		Bukkit.getPluginManager().registerEvents(this, plugin);
	}
	
	@Override
	public void load(){
		Bukkit.getScheduler().runTask(plugin, () -> Bukkit.getOnlinePlayers().forEach(this::checkPlayer));
	}
	
	@Override
	public void unload(){
		for (Player p : Bukkit.getOnlinePlayers()) {
			for (Player pl : Bukkit.getOnlinePlayers()) {
				p.showPlayer(pl);
			}
		}
		HandlerList.unregisterAll(this);
	}
	
	@EventHandler
	public void onJoin(PlayerJoinEvent e) {
		long time = System.nanoTime();
		checkPlayer(e.getPlayer());
		TAB.getInstance().getCPUManager().addTime(TabFeature.PER_WORLD_PLAYERLIST, UsageType.PLAYER_JOIN_EVENT, System.nanoTime()-time);
	}
	
	@EventHandler
	public void onWorldChange(PlayerChangedWorldEvent e) {
		long time = System.nanoTime();
		checkPlayer(e.getPlayer());
		TAB.getInstance().getCPUManager().addTime(TabFeature.PER_WORLD_PLAYERLIST, UsageType.WORLD_SWITCH_EVENT, System.nanoTime()-time);
	}
	
	private void checkPlayer(Player p) {
		for (Player all : Bukkit.getOnlinePlayers()){
			if (all == p) continue;
			if (!shouldSee(p, all) && p.canSee(all)) p.hidePlayer(all);
			if (shouldSee(p, all) && !p.canSee(all)) p.showPlayer(all);
			if (!shouldSee(all, p) && all.canSee(p)) all.hidePlayer(p);
			if (shouldSee(all, p) && !all.canSee(p)) all.showPlayer(p);
		}
	}
	
	private boolean shouldSee(Player viewer, Player target) {
		if (target == viewer) return true;
		if ((allowBypass && viewer.hasPermission("tab.bypass")) || ignoredWorlds.contains(viewer.getWorld().getName())) return true;
		String viewerWorldGroup = viewer.getWorld().getName() + "-default"; //preventing unwanted behavior when some group is called exactly like a world
		String targetWorldGroup = target.getWorld().getName() + "-default";
		for (Entry<String, List<String>> group : sharedWorlds.entrySet()) {
			if (group.getValue() != null) {
				if (group.getValue().contains(viewer.getWorld().getName())) viewerWorldGroup = group.getKey();
				if (group.getValue().contains(target.getWorld().getName())) targetWorldGroup = group.getKey();
			}
		}
		return viewerWorldGroup.equals(targetWorldGroup);
	}
}

File: krypton/src/main/kotlin/me/neznamy/tab/platforms/krypton/features/PerWorldPlayerList.kt
package me.neznamy.tab.platforms.krypton.features

import me.neznamy.tab.api.TabFeature
import me.neznamy.tab.platforms.krypton.Main
import me.neznamy.tab.shared.TabConstants
import me.neznamy.tab.shared.TAB
import org.kryptonmc.api.entity.player.Player
import org.kryptonmc.api.event.Listener
import org.kryptonmc.api.event.player.JoinEvent

// TODO: Also check player on world change when Krypton supports world changing
class PerWorldPlayerList(private val plugin: Main) : TabFeature("Per world playerlist", null) {

    private val allowBypass = TAB.getInstance().configuration.config.getBoolean(
        "per-world-playerlist.allow-bypass-permission",
        false
    )
    private val ignoredWorlds = TAB.getInstance().configuration.config.getStringList(
        "per-world-playerlist.ignore-effect-in-worlds",
        listOf("ignoredworld", "build")
    )
    private val sharedWorlds = TAB.getInstance().configuration.config
        .getConfigurationSection<String, List<String>>("per-world-playerlist.shared-playerlist-world-groups")

    init {
        TAB.getInstance().debug("Loaded PerWorldPlayerList feature with parameters allowBypass=$allowBypass, ignoredWorlds=$ignoredWorlds, " +
            "sharedWorlds=$sharedWorlds")
        plugin.server.eventManager.register(plugin, this)
    }

    @Listener
    fun onJoin(event: JoinEvent) {
        val time = System.nanoTime()
        checkPlayer(event.player)
        TAB.getInstance().cpuManager.addTime(featureName, TabConstants.CpuUsageCategory.PLAYER_JOIN, System.nanoTime() - time)
    }

    override fun load() {
        plugin.server.scheduler.run(plugin) { plugin.server.players.forEach(::checkPlayer) }
    }

    override fun unload() {
        plugin.server.players.forEach { plugin.server.players.forEach(it::show) }
        plugin.server.eventManager.unregisterListener(plugin, this)
    }

    private fun checkPlayer(player: Player) {
        plugin.server.players.forEach {
            if (it === player) return@forEach
            if (!shouldSee(player, it) && player.canSee(it)) player.hide(it)
            if (shouldSee(player, it) && !player.canSee(it)) player.show(it)
            if (!shouldSee(it, player) && it.canSee(player)) it.hide(player)
            if (shouldSee(it, player) && !it.canSee(player)) it.show(player)
        }
    }

    private fun shouldSee(viewer: Player, target: Player): Boolean {
        if (target === viewer) return true
        if ((allowBypass && viewer.hasPermission(TabConstants.Permission.PER_WORLD_PLAYERLIST_BYPASS)) || ignoredWorlds.contains(viewer.world.name)) {
            return true
        }
        var viewerWorldGroup = "${viewer.world.name}-default"
        var targetWorldGroup = "${target.world.name}-default"
        sharedWorlds.forEach { (key, value) ->
            if (value == null) return@forEach
            if (value.contains(viewer.world.name)) viewerWorldGroup = key
            if (value.contains(target.world.name)) targetWorldGroup = key
        }
        return viewerWorldGroup == targetWorldGroup
    }
}
